### ISSUE TITLE
Usage in Angular Universal 

### ISSUE THREAD

**Issue by AidasMIT** [damienbod/angular-auth-oidc-client#1277]
Hello guys, I faced very strange issue with Angular Universal. 
I get ERROR Error: Uncaught (in promise): [object Undefined]

**To Reproduce**
1. Install new angular app, add your angular auth oidc client to the project. 
2. In App.module.ts add 
```
  AuthModule.forRoot({
      config: {
        authority: '<your authority address here>',
        redirectUrl: 'setrandomurl',
        postLogoutRedirectUri: 'setrandomurl',
        clientId: '<your clientId>',
        scope: 'openid profile email offline_access',
        responseType: 'code',
        silentRenew: true,
        useRefreshToken: true,
        logLevel: LogLevel.Debug,
      },
    }),
```

**Expected behavior**
Then add this to your app.module.ts file and run command -> npm run dev:ssr 
I get this error mention before, could you please tell why is that happened?

**Comment by CesarD** [comment]
Does this library actually works with Angular Universal?
I'm an old user of `oidc-client` and later `oidc-client-ts` for Angular (SPA), but now I'm researching to use Angular Universal and I would like to get an idea of what libraries work on it and if the configurations change at all between both scenarios: SPA and SSR. 
Thanks.

**Comment by gourav-sparkt** [comment]
@CesarD 
**Hey sorry for pinging you directly!**

Did you get solution for this SSR issue?
any new library?
We are developing a blog website and if SSR does not work then we need to **abandon Angular** :(

**Comment by CesarD** [comment]
> @CesarD **Hey sorry for pinging you directly!**
> 
> Did you get solution for this SSR issue? any new library? We are developing a blog website and if SSR does not work then we need to **abandon Angular** :(

Hey there.
I have been able to implement it for an SPA and integrate it there, but right now I'm still experimenting and trying to find my way around Angular Universal... I have been able to perform the login of a user on client-side, but despite having installed ngx-cookie and ngx-cookie-backend for providing support for cookies on both client and server-side, the library doesn't yet recognizes my custom storage service based on cookies and doesn't work server-side.

It would be great to have much more docs and samples as all that is very scarce from what I've seen around...

I'm on the same boat as you, as I'd like to be able to port an app I have as experiment of the feasibility of Angular Universal for other projects, but we need much more detail of many subtleties that come with Angular Universal.

**Comment by gourav-sparkt** [comment]
> > @CesarD **Hey sorry for pinging you directly!**
> > Did you get solution for this SSR issue? any new library? We are developing a blog website and if SSR does not work then we need to **abandon Angular** :(
> 
> Hey there. I have been able to implement it for an SPA and integrate it there, but right now I'm still experimenting and trying to find my way around Angular Universal... I have been able to perform the login of a user on client-side, but despite having installed ngx-cookie and ngx-cookie-backend for providing support for cookies on both client and server-side, the library doesn't yet recognizes my custom storage service based on cookies and doesn't work server-side.
> 
> It would be great to have much more docs and samples as all that is very scarce from what I've seen around...
> 
> I'm on the same boat as you, as I'd like to be able to port an app I have as experiment of the feasibility of Angular Universal for other projects, but we need much more detail of many subtleties that come with Angular Universal.

@CesarD thanks! I think we need to move to nextJs or something for blog websites. I
will keep using angular for admin consoles <3.

**Comment by CesarD** [comment]
I'd try to experiment a bit more and try to work my way around it.
I've read enough to understand that it's possible, but unfortunately the docs are a bit scarce regarding this part of Angular so it's harder to understand some things.
Try with the Discord server, there's many people there willing to answer questions and help out 😉

**Comment by CesarD** [comment]
@gourav-sparkt I have been asking and experimenting a little bit more about how Angular Universal works and from what I learned, the idea is that the first page that gets rendered when you enter through any URL to your app, is the one that gets rendered on the server and served ready to be displayed in the browser. This helps SEO in the way that if a web crawler or any other bot sends a request to your site, the server will return a fully rendered page which can be correctly indexed without executing any JS. From there, and through the links available through anchors `<a>`, the bot/crawler can navigate your site structure as any other site, and any other request will be served the same way I explained above, even though any user accessing through a browser will only get the first page server-side rendered and, afterwards, the rest of the bundle will be loaded and the app will behave just like any other SPA...
Of course, all this SSR part should occur under the assumption that these pages will be accessible without any authentication, otherwise, protecting a page that should be anonymously accessible by an unauthenticated bot/crawler wouldn't make much sense... So the login and authentication shouldn't be needed at all on this scenario and I have been able to make it work very rudimentarily without throwing any exception, although some information errors about the lib not recognizing the storage I set on the server.

I am now in the process of setting things up to build another PoC that covers the basic stuff a bit more decently than my first experiment.

**Comment by sparktindia** [comment]
@CesarD That should actually work! I'll try to do the same. Please do let me know what insights you find further in the process

**Comment by gourav-sparkt** [comment]
@CesarD That should work! I'll try to do the same. Please do let me know what insights you find further in the process

**Comment by CesarD** [comment]
@gourav-sparkt Just wanted to come back to mention that I've been implementing this sample application and that everything seems to be working as expected: the SSR part is rendering the page always anonymously and the hydration helps to render the client-side rendered part correctly on top of it and the remaining of the pages work normally as expected. The login works correctly for the SPA part of the app and other than some server-side information logs, at least there don't seem to be errors being logged in the console, so I take that as success 😄 
Also, I had to reconsider some things to make the app truly SEO friendly, like ie: if I have a catalog of products for the user to browse, I had to create an alternative version of the Bootstrap pagination control for it to render the button with actual links and not just buttons that only work client-side, otherwise a crawler/bot won't be able to traverse my site structure and get to the rest of the products because the page wouldn't be retrieved with actual links rendered server-side. But this is completely aside of the authentication stuff we were talking before.
Hope you get this to work for you as well to keep everything under Angular 😉
